### PR TITLE
Merge scatter and gather indexing implementations.

### LINKED ISSUE #1006: Load from 2d array within vmap results in two gathers rather than just one
```
import jax
import jax.numpy as np
import numpy as onp

def test(arr, x, y):
  return arr[x, y]

rand_x = onp.random.randint(0, 200, size=1024)
rand_y = onp.random.randint(0, 200, size=1024)
rand_arr = onp.random.uniform(0, 1, size=(200,200))
test_vmap = jax.vmap(test, in_axes=(None, 0, 0))
jax.jit(test_vmap)(rand_arr, rand_x, rand_y)
```

The `jax.jit` results in an XLA computation with two gathers, but the expression can be implemented with just one.

Two gathers is quite bad, because unnecessarily it materializes out a large array; you really don't want to do this.

I am not blocked on this, because I've worked around it by flattening the array and doing `x * width + y`.  But this is not ideal because it forces a layout upon XLA.  Also this is something that nobody other than an XLA expert is going to notice and work around in their code.

### PR DESCRIPTION
**Pull request by hawkinsp**
I would still like to do some performance benchmarking (i.e., write some benchmarks), but sending this out for an initial review.

Fixes #1006

### CODE REVIEW

File: tests/lax_numpy_indexing_test.py [hunk 1]
```diff
@@ -541,6 +541,7 @@ def fun(unpacked_indexer, x):
       for dtype in all_dtypes
       for rng in [jtu.rand_default()])
   def testAdvancedIntegerIndexing(self, shape, dtype, rng, indexer):
+    print("indexer", indexer)
```
Comment (mattjj) [inline tests/lax_numpy_indexing_test.py]: Want to keep this line in?

File: jax/ops/scatter.py [hunk 2]
```diff
@@ -64,178 +45,32 @@ def _scatter_update(x, idx, y, scatter_op):
...
-    y = lax.rev(y, reversed_y_dims)
 
+  # XLA gathers and scatters are very similar in structure; the scatter logic
+  # is more or less a transpose of the gather equivalent.
+  indexer = np._index_to_gather(np.shape(x), idx)
+
+  # Broadcast `y` to the slice output shape.
+  y = np.broadcast_to(y, tuple(indexer.slice_shape))
+  # Collapse any `None`/`np.newaxis` dimensions.
+  #y = lax.reshape(y, indexer.collapsed_slice_shape)
```
Comment (mattjj) [inline jax/ops/scatter.py]: Extra line?

**Review by mattjj — APPROVED** [review]
Awesome, this is a great improvement by generating only a single Gather HLO. It's hard to check the logic in detail but that's why we have tests :)

We spoke offline and discussed how we'd really like to have performance benchmarks to measure against (e.g. by not generating lax.slice calls in special cases, could this code be less performant in those special cases?) but in the absence of those we can probably rely on our users to tell us when they need indexing to be faster and we can add special cases as needs arise.

File: tests/lax_numpy_indexing_test.py (hunk 1 above)
Comment (hawkinsp) [reply tests/lax_numpy_indexing_test.py]: Removed.

File: jax/ops/scatter.py (hunk 2 above)
Comment (hawkinsp) [reply jax/ops/scatter.py]: Deleted.

**Pull request merged**